test/main.cpp

```diff
@@ -34,44 +39,126 @@
+    result += buffer.data();
+  }
+  return result;
+}
+
+// manifestfile must be in abs path format
+static void read_tests_from_manifest(string manifestfile, set<string>& run_list,
+                                     set<string>& skip_list) {
+  static set<string> g_imported_files;
   fstream fs;
-  fs.open(filter_file, ios::in);
+  fs.open(manifestfile, ios::in);
   if (fs.is_open()) {
-    std::string filters = "";
-    std::string line;
-    bool excluded_section = false;
+    cout << "Parsing manifest: " << manifestfile << " ...\n";
+    string line;
+    string curr_section = "";
+    g_imported_files.insert(manifestfile);
     while (getline(fs, line)) {
       line = std::regex_replace(line, std::regex("^\\s+"), "");
       line = std::regex_replace(line, std::regex("#.*$"), "");
       line = std::regex_replace(line, std::regex("\\s+$"), "");
       if (line.empty()) continue;
-      if (!excluded_section && line.find("[EXCLUDED]") == 0) {
-        if (filters.back() == ':') filters.pop_back();  // remove last :
-        filters += "-";
-        excluded_section = true;
+      if (std::regex_search(line, std::regex("^\\[IMPORT\\]$"))) {
+        curr_section = "import_section";
+        continue;
+      }
+      if (std::regex_search(line, std::regex("^\\[RUN\\]$"))) {
+        curr_section = "run_section";
+        continue;
+      }
+      if (std::regex_search(line, std::regex("^\\[SKIP\\]$"))) {
+        curr_section = "skip_section";
+        continue;
+      }
+      if (curr_section == "import_section") {
+        if (g_imported_files.find(line) != g_imported_files.end()) {
+          cout << "ERROR: re-import of manifest " << line << " in "
+               << manifestfile << endl;
+          exit(1);
+        }
+        line = SCRIPTDIR + "/" + line;
+        g_imported_files.insert(line);
+        set<string> new_runs, new_skips;
+        read_tests_from_manifest(line, new_runs, new_skips);
+        run_list.insert(new_runs.begin(), new_runs.end());
+        skip_list.insert(new_skips.begin(), new_skips.end());
+      }
+      if (std::regex_search(line, std::regex("[:\\s]"))) {
+        cout << "Bad pattern: [" << line << "], ignoring...\n";
         continue;
       }
-      if (excluded_section) {
-        if (line.at(0) == '-') line = line.erase(0, 1);
+      if (curr_section == "run_section") {
+        run_list.insert(line);
+      }
+      if (curr_section == "skip_section") {
+        skip_list.insert(line);
       }
-      filters += line + ":";
     }
     fs.close();
-    if (filters.back() == ':') filters.pop_back();  // remove last :
+  } else {
+    cout << "Cannot open file: <" << manifestfile << ">\n";
+  }
+}
 
-    ::testing::GTEST_FLAG(filter) = filters;
+// To keep the logic in one place, we will invoke function from test_utils.py
+static string get_test_manifest_filepath() {
+  string cmdstr =
+      "python3 " + SCRIPTDIR +
+      "/../tools/test_utils.py --run_func print_test_manifest_filename";
```

Comment:
Would we rather prefer to have a duplicate implementation locally in C++?

Reply:
Yes.